fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraResourceImpl.java

```diff
@@ -774,6 +800,62 @@ public Node getNodeVersion(final String label) {
 
     }
 
+    @Override
+    public String getVersionLabelOfFrozenResource() {
+        if (!isFrozenResource()) {
+            return null;
+        }
+
+        // Version History associated with this resource
+        final VersionHistory versionHistory = getUnfrozenResource().getVersionHistory();
+
+        // Frozen node is required to find associated version label
+        final Node frozenResource;
+        try {
+            // Possibly the frozen node is nested inside of current child-version-history
+            if (getNode().hasProperty(JCR_CHILD_VERSION_HISTORY)) {
+                final Node childVersionHistory = getNodeByProperty(getProperty(JCR_CHILD_VERSION_HISTORY));
+                final Node childNode = getNodeByProperty(childVersionHistory.getProperty(JCR_VERSIONABLE_UUID));
+                final Version childVersion = getVersionManager().getBaseVersion(childNode.getPath());
+                frozenResource = childVersion.getFrozenNode();
+
+            } else {
+                frozenResource = getNode();
+            }
+
+            // Loop versions
+            final VersionIterator versions = versionHistory.getAllVersions();
```

Comment:
Use `Iterator<Version>` here at the cost of a known-good typecheck and you can use `forEachRemaining` which is shorter.